src/compiler/checker.ts

```diff
@@ -5503,6 +5503,10 @@ namespace ts {
                 return symbol.declarations && find(symbol.declarations, s => !!getEffectiveTypeAnnotationNode(s) && (!enclosingDeclaration || !!findAncestor(s, n => n === enclosingDeclaration)));
             }
 
+            function getExistingNodeHasNoTypeParametersOrMatchingTypeParameters(existing: TypeNode, type: Type) {
+                return !(getObjectFlags(type) & ObjectFlags.Reference) || !isTypeReferenceNode(existing) || length(existing.typeArguments) >= getMinTypeArgumentCount((type as TypeReference).target.typeParameters);

```

Comment:
This is kinda surprising to me - this checks whether the existing type has *more* type arguments which I'd assume to be the opposite condition.

Reply:
Yeah, if the existing declaration provides the minimum required or more, it's good to reuse; how's that surprising?

Reply:
Maybe it's a naming issue? What is the meaning of `hasNoTypeParameters` in this case?